devto draft (#1)
vincentntang/vincent-blog-draft: devto-draft.md

---
title: Learning Vue as a React Developer
published: false
description:
tags:
---

> React. The first major javascript framework I decided to learn to come from jQuery land.

Reviewer: ```suggestion
React. The first major JavaScript framework I decided to learn coming from jQuery land.
```

> My journey learning was painful. I had to watch 3 different Udemy tutorials to finally get it. Months were spent wrapping my head around this foreign way of coding.

Reviewer: ```suggestion
My journey learning it was painful. I had to watch three different Udemy tutorials to finally get it. Months were spent wrapping my head around this foreign way of coding.
```

> I slowly and surely learned the basics. On how state management worked. Or how the React lifecycle operated. These things came with time, and eventually, I learned to use this knowledge to build apps featured in hackathons. And things I would use in my daily life.

Reviewer: ```suggestion
I slowly and surely learned the basics, such how state management works and how the React lifecycle operates. These things came with time and eventually I learned to use this knowledge to build apps featured in hackathons and to build tools I would use in my daily life.
```

> Eventually, I started my first development job. They used Vue. I didn't know anything about this framework, and it felt like I had to learn javascript all over again.

Reviewer: ```suggestion
The time came to start my first development job as part of a team that used Vue. I didn't know anything about this framework and I felt like I had to learn JavaScript all over again.
```

> These are the things that I had to learn, transitioning from React to Vue

Reviewer: ```suggestion
This intense but interesting new journey where I transitioned from React to Vue left me with the following valuable lessons that I want to share with you.
```

## Styling and Syntax

> Transitioning from React to Vue was liberating. It was like revisiting the good old days of EJS or any templating language. No longer did I have to write my HTML, and then convert it to JSX. I could just write HTML^TM and be on my way.

Reviewer: Define any acronym when first introduced. I think in this case:

>  EJS (Embedded JavaScript templates)

Here's an example of what a typical Vue file looks like:

```vue
<template>
  <div>
    <h1>{{ message }}</h1>
  </div>
</template>

<script>
export default {
  data() {
    return {
      message: "Hello World"
    };
  }
};
</script>

<style lang="scss" scoped>
.h1 {
  color: red;
}
</style>
```

> Assuming you declared your Vue component at the top level, this just outputs a "Hello World" in red color.

Reviewer: ```suggestion
Assuming you declared your Vue component at the top level, this just outputs a "Hello World" string in red color.
```

> In Vue, you write your HTML, javascript in the center, and styles at the bottom. This opinionated logic coupling makes scaling Vue apps easy.

Reviewer: ```suggestion
In Vue, you write your HTML, JavaScript in the center, and styles at the bottom. This opinionated logic coupling makes scaling Vue apps easy.
```

> In Vue, you write your HTML, javascript in the center, and styles at the bottom. This opinionated logic coupling makes scaling Vue apps easy.

Reviewer: I recommend a different structure and message here:

> In Vue, you write single file components. By convention, your component file will include the declaration of your template at the top, followed by a script definition with all the component logic, and it wii close with any necessary style declarations.

> React, on the other hand, is less opinionated on what you do with styles. You can use CSS-in-JS approach like this:

Reviewer: Suggestion:

> React, on the other hand, is less opinionated on file structure. You can still create single file components but instead of implementing the template in HTML and the styling in SCSS, for example, you would write  everything in Javascript: template in JSX and styling in CSS-in-JS. This is demonstrated in this example:

```js
import React, { Component } from "react";

export default class MyFirstComponent extends Component {
  state = {
    message: "Hello World"
  };
  render() {
    return (
      <div>

>         <h1 style={styles.color}>{this.state.message}</h1>

Reviewer: I feel this should be `styles.header.color`

      </div>
    );
  }
}

// CSS in JS
const styles = {
  header: {
    color: "red"
  }
};
```

> but you could also just import a `.css` or `.scss` file as well. The problem you run into though when using an import, is you pollute the global namespace. As your app and your team grows, chances are high your developer co-worker might name a class the same as yours.

Reviewer: ```suggestion
But you could also just import a `.css` or `.scss` file. However, one of the problems that you can run into when importing CSS files is the pollution of the global namespace. As your app and your team grows, chances are high that your teammates might name a class the same as yours and create a collision bug.
```

> You could opt to use a library like `styled-components`. This gives you a Sass-like syntax to tightly couple your styles with your JSX.

Reviewer: ```suggestion
You could opt to use a library like `styled-components` that gives you a Sass-like syntax to tightly couple your styles with your JSX.
```

> React brings another set of issues. Do you use a functional component or stateful component? By default, it's better to use only what you need, but this means potentially more refactoring down the road. This doesn't factor in that you can use React Hooks now instead of class components.

Reviewer: This seems to be a tangent. It's not explicitly related to the topic of this section which is file structure. You could make this another section if it's important to you.

## Props, State

> React has something called props, which is data loaded in from the parent component. And something called state, data intrinsic to that component.

Reviewer: Use this document as source for the definition of props. 

https://reactjs.org/docs/components-and-props.html

I think it's ideal to focus on the value of props as artifacts for one-way data flow.

Here's an example of just rendering a "Hello World" message:

```js
// React Parent Component
import React, { Component } from "react";
import Child from "./Child";

export default class Parent extends Component {
  state = {
    message: "Hello World"
  };
  render() {
    return (
      <div>
        <Child message={this.state.message} />
      </div>
    );
  }
}
```

```js
// React Child Component
import React from "react";

const Child = props => {
  return <div>{props.message}</div>;
};
export default Child;
```

> With Vue, the concepts were the same. But Vue prefers convention over configuration. Meaning, there are specific naming conventions. Props being passed from parent component downward are in lowercase-hyphens. When the prop is declared from the child level, it is in kebab case.

Reviewer: ```suggestion
With Vue, the concepts are the same. But Vue embraces convention over configuration. This is expressed by having specific naming conventions, props being passed downward from a parent component must be in lowercase-hyphens. When the prop is declared from the child level, it is in kebab case.
```

Reviewer: I think "lowercase-hyphens" and "kebab-case" are the same. Double check that as I am not familiar with Vue idioms.

Example:

> ```vue

Reviewer: I don't see how this example demonstrated the named props contrast explained in this section.

<!-- Vue Parent Component -->
<template>
  <div>
    <child-component :message="message" />
  </div>
</template>

<script>
import ChildComponent from "./ChildComponent.vue";
export default {
  data() {
    return {
      message: "hello world"
    };
  },
  components: {
    ChildComponent
  }
};
</script>
```

```vue
<!-- Vue Child Component -->
<template>
  <div>
    {{ message }}
  </div>
</template>

<script>
export default {
  props: {
    message: {
      type: String,
      default: undefined
    }
  }
};
</script>
```

> Vue requires you to declare a child component twice, once on import another on `Components`.

Reviewer: This is another part that feels tangential as the section is named "Props and State" perhaps that belong to the section talking about file structure.

React differs in that there is one less step, you import the file then use it.

## Type Enforcement

> One thing I miss in static languages like C# or Java is type enforcement. Javascript is a dynamic scripting language. Meaning it doesn't care if a variable is a string, number, boolean, etc.

Reviewer: ```suggestion
One thing I miss from static languages like C# or Java is type enforcement. JavaScript is a dynamic scripting language so it doesn't care if a variable is a string, number, boolean, etc.
```

> One thing I miss in static languages like C# or Java is type enforcement. Javascript is a dynamic scripting language. Meaning it doesn't care if a variable is a string, number, boolean, etc.

Reviewer: > it doesn't care if a variable is a string, number, boolean, etc.

I am not too sure about that. It does care because using the wrong type of variable can break an app, it just doesn't enforce it. It's weakly typed but the types do exist.

> When you declare a string, you can always typecast it to a number after. This makes scaling large frontend apps difficult. Because you might not know what data type you're working with.

Reviewer: Depends on the string. You cannot typecast "github" to a number. 

```
let string = "github";
let number = parseInt(text);
// number = NaN
```

> React natively solves this with `PropTypes`. Example in the `<Child/>` component we made earlier:

Reviewer: ```suggestion
React natively solves this with `PropTypes`. Check out this example based  on the `<Child/>` component that was created earlier:
```

```js
import React from "react";
import PropTypes from "prop-types";

const Child = props => {
  return <div>{props.message}</div>;
};

Child.propTypes = {
  message: PropTypes.string.isRequired
};

export default Child;
```

There are a few issues with propTypes in React

- It's really easy to misspell
- It's optional

> - You have to declare an import

Reviewer: > You have to import a module

I think this sounds better.

> Because of these reasons, I find myself forgetting to use propTypes. You could opt to use Typescript instead with React, but this just means more additional time configuring your setup.

Reviewer: I recommend expanding why those are problematic things about PropTypes in React, specially the import one.
Also check the spelling to consistently be "PropTypes".

> Because of these reasons, I find myself forgetting to use propTypes. You could opt to use Typescript instead with React, but this just means more additional time configuring your setup.

Reviewer: ```suggestion
Because of these reasons, I find myself forgetting to use PropTypes. You could opt to use Typescript instead with React, but this just means additional overhead configuring your setup.
```

> Vue forces you to declare your props, in a way that makes sense. Inside your `<script>` tag where all the other logic lives

Reviewer: ```suggestion
Vue forces you to declare your props in a way that makes sense: Inside your `<script>` tag where all the other logic lives.
```

> Vue forces you to declare your props, in a way that makes sense. Inside your `<script>` tag where all the other logic lives

Reviewer: This should be unpacked: Why does it make sense? What's good about it? Also mention that the props are typed when declared.

> Example for a Vue child component:

Reviewer: ```suggestion
Example featuring a Vue child component:
```

```vue
<template>
  <div>
    {{ message }}
  </div>
</template>

<script>
export default {
  props: {
    message: {
      type: String,
      default: undefined
    }
  }
};
</script>

<style lang="scss" scoped></style>
```

> This is a much better design overall and makes it so that Typescript is not necessary in Vue.

Reviewer: See above. This is a critic that needs unpacking and I recommend adding more depth to this conclusion as it opinionated as mostly everything in the domain of frameworks.

## Data Reactivity

> This was the concept that tripped me on Vue several times. In React, everything is reactive. Meaning you can add properties to objects, and React would call it's lifecycle when properties changed.

Reviewer: ```suggestion
This was a concept that tripped me on Vue several times. In React, everything is reactive. Meaning you can add properties to objects, and React would call its lifecycle when properties changed.
```

> This was the concept that tripped me on Vue several times. In React, everything is reactive. Meaning you can add properties to objects, and React would call it's lifecycle when properties changed.

Reviewer: Some say React is not reactive:
https://www.reddit.com/r/reactjs/comments/98lbpi/why_react_is_not_reactive_react_rally_2018/

> In Vue ... things don't work the same. In React, every time you modify state, you could pass a new entire state object.

Reviewer: ```suggestion
In Vue, things work differently. In React, every time you modify state, you could pass a new entire state object.
```

> Vue differed in that you _mutated_ your state directly through methods. Because we're intrinsically adding side effects, you have to actually _declare_ when new properties are being added to the object.

Reviewer: ```suggestion
In Vue, you _mutate_ your state directly through methods. Because you're intrinsically adding side effects, you have to actually _declare_ when new properties are being added to the object.
```

Ensure that pronounces are always on check and be consistent with their use. Here you were using "we" and "you" in the same sentence.

> Namely using something called `Vue.set()`.

Reviewer: This is a dangling sentence. Attach it to the previous paragraph or build it up by adding more context to it.

> A common example of when you would use this is if your loading data from the backend. And need to create new data to send back.

Reviewer: ```suggestion
A common example of when you would use this is if your loading data from the backend and need to create new data to send back.
```

Again, this is dangling. Collect your sentences into paragraphs and avoid using clauses.

> Another concept is if you

Reviewer: Missing text?

## Key Values

> Whenever you iterate through and create several child components, Vue/React need to uniquely identify each item.

Reviewer: This could be a single sentence that's part of the "Props, State" section

With Vue, you use the `:key` and pass a value. In React it's more or less the same thing

## Watchers and Observables

> I never quite understood design patterns that well. But the term "observable" finally made sense to me.

Reviewer: > I never quite understood design patterns that well

Do you mean to express design patterns in general? I think it's safe to leave that out and just say that you were able understand the observable design pattern through time or something like that.

> Vue has something called watchers. This lets you call a getter everytime that variable changes.

Reviewer: ```suggestion
Vue has something called watchers that let you call a getter every time a variable changes.
```

> Vue has something called watchers. This lets you call a getter everytime that variable changes.

Reviewer: Collect these sentences into a paragraph.

However, there's some caveats, because you could have watchers watching watchers. This causes circular referential issues.

The Vue docs warn against this. I myself have accidentally done this using inputs to watch over other inputs.

> React does not have anything similar out of the box.

Reviewer: I think I created this problem using Hooks and the `useEffect method.

## Computed properties

> This was a new concept to me, from the React world. In Vue, you can have variables that depend on other variables. Anytime a dependency changes, so does the computed property.

Reviewer: ```suggestion
As a React developer, this was a new concept to me. In Vue, you can have variables that depend on other variables. Anytime a dependency changes, so does the computed property.
```

It seems useful at first, but computed properties are only good if you aren't modifying it at all directly.

> When you start needing to, this is when you go away from computed properties entirely to the next item...

Reviewer: Connect this sentence with the one before.

## Methods

Methods in Vue worked the same as in React. These methods were called through event-directives that extend from HTML DOM events. Examples include putting `@click` events, or `@input` if a user typed text in an input field.

Methods are what you'll always fall back in Vue when computed properties and watchers don't fit your use-case.

> Methods are like any function in javascript. The return statement is totally optional, you can just use it to mutate data in the background _(not a good practice)_

Reviewer: Add final period to end the paragraph or sentence.

## Asynchronous Gotchas

> This is a gotcha that you find out later in Vue, and in React.

Reviewer: ```suggestion
This is a gotcha that you find out later in Vue and in React.
```

In React, there is a built-in method called `setState()`. When `setState()` gets called, it gets handled asynchronously in the background.

> Vue is no different, you have a method called `Vue.nextTick()` that awaits the latest updated value to the data.

Reviewer: To the data or to the state?

> One lesson I learned is that it's best to set local variables within your methods, and only mutate data attributes as needed.

Reviewer: ```suggestion
One lesson I learned is that it's best to set local variables within your methods and only mutate data attributes as needed.
```

## Templates

With Vue, there is something called `templates`. In your component, you have the ability to declare `<template>` slots. When that component gets called, you can inject HTML where those slots match.

> This has been immensely useful when working in libraries. I could inject functionality as I saw fit, to get the feature I needed.

Reviewer: What do you mean by "working in libraries"?

> This has been immensely useful when working in libraries. I could inject functionality as I saw fit, to get the feature I needed.

Reviewer: > as I saw fit, to get the 

No comma

Templates do not get rendered as a `<div>` element in the DOM. React has a similar approach to templates called fragments, but the approach is limited to just the first element in your component.

## Lifecycle methods

Lifecycle methods describe what components do during their lifetime.

Here are common ones you'll use in Vue:

> 

Reviewer: All component names need to be fenced in Markdown: `componentDidMount`, etc.

> - mounted() - Similar to React's componentDidMount. This generally when making an Axios/fetch call to initialize data in your component from a backend.

Reviewer: The sentence structure is not there. 

> This generally when making an Axios/fetch call to initialize data in your component from a backend.

Maybe: This method is triggered when making AJAX calls to initialize data in your component from a backend.

> - updated() - Similar to React's componentDidUpdate. When you update data through your components, you want to send a POST request to your backend to keep things in sync

Reviewer: Missing period at the end.

- destroyed() - Similar to React's componentWillUnmount. You don't need to use this, but it helps clean up leftover event listeners.

## Hooks

> Hooks makes React an extremely powerful framework. No longer did you have to use redux for handling state, you could just use `useState()` instead. No longer did you have to deal with `setState()`, there were a slew of other new improved methods to use.

Reviewer: React is not technically a framework. Hooks are a library enhancement.

> Hooks makes React an extremely powerful framework. No longer did you have to use redux for handling state, you could just use `useState()` instead. No longer did you have to deal with `setState()`, there were a slew of other new improved methods to use.

Reviewer: The form is not there for this sentence:

> No longer did you have to use redux for handling state, you could just use `useState()` instead. 

Maybe: You can handle state easily by using the `useState` hook without having to resort to tools like Redux

> Vue has no equivalent to hooks, although there is an RFC at the time of writing this

Reviewer: Missing period at the end.

## Directives

Vue's built-in directives make development a breeze. You could create iterable directives right inside your HTML.

> In React, this often meant making another component called "group(s)" and maybe one called "group".

Reviewer: I don't understand what is being said there tbh.

> Vue, you can just dump the `v-for`, `v-if` etc right inside your HTML. Your logic coherently just makes sense looking at it.

Reviewer: > In Vue,

> Vue, you can just dump the `v-for`, `v-if` etc right inside your HTML. Your logic coherently just makes sense looking at it.

Reviewer: This sentence form needs improvement. 

> Vue, you can just use directives like `v-for` and `v-if` right inside your HTML. Their names and position let you  make sense of the template logic quickly.

React pretty much does the same but it uses JavaScript instead of a DSL

> Alternatively, you can write your own directives! This lets you attach functionality to page if necessary, making it easier to develop faster

Reviewer: ```suggestion
Alternatively, you can write your own directives! This lets you encapsulate and share functionality across components easily when necessary, helping you work faster.
```

> One great use case for startups is to develop a reporting system. If a user clicks on a feature that doesn't yet exist, you can add a popup modal saying "Sorry we're working hard on this new feature! Come back later". In the background, you can make Axios request to notify users really do want that feature.

Reviewer: How is this a showcase of directive? I suggest unpacking more how this is an example of a directive and perhaps show a code example.

## Mixins

> Some companies swear by these. It's a reusable code snippet library that can be implemented across multiple components. React out of the box does not have such a feature, you can alternatively use callback functions instead that are made globally available

Reviewer: I wouldn't use "companies" here. Developers is a better term.

## Emitting Events

Props are passed from parent to child component, via one-way data binding. To handle business logic at the parent level, from the child component, you generally emit events.

With Vue, there are two different ways.

> 1. You can pass a reference to the function from parent to child

Reviewer: ```suggestion
1. Pass a reference to the function from parent to child
```

> 2. Emit an event from the child and capture the response in parent

Reviewer: ```suggestion
2. Emit an event from the child and capture the response in the parent
```

> Most applications generally use #2. Example from a button that triggers a method in the parent component

Reviewer: ```suggestion
Most applications generally use the second approach. This is an example from a button that triggers a method in the parent component:
```

```vue
<!-- Parent Component -->
<template>
  <ChildComponent @event-emitted="_handleUpdatedEvent" />
</template>

<script>
import ChildComponent from "./components/ChildComponent.vue";

export default {
  components: {
    ChildComponent
  },
  methods: {
    _handleUpdatedEvent(val) {
      console.log("Button was clicked");
    }
  }
};
</script>
```

```vue
<!-- Child Component -->
<template>
  <div class="hello">
    <button @click="$emit('event-emitted')">Click Me</button>
  </div>
</template>
```

## Global State Management

> Vue's CLI comes with the ability to add Vuex out of the box.

Reviewer: Collect the sentences in a paragraph. I cannot really comment on the technical part of this as it's not my domain.

In React you pass a new set of state using `setState()`, Redux only extends this through global dispatches/flux architecture.

In Vue, you mutate objects behind the scene. Vuex is no different but at the global level.

Vue natively ships with the ability to create a global event bus, which is similar to React's context API.

## Final Thoughts

Learning Vue has been a great way to learn new programming paradigms.

Many principles in React translate over to Vue. There are a few things that differ, namely:

> - You don't replace your state, you mutate it. There's data reactivity / async gotchas from these

Reviewer: ```suggestion
- You don't replace your state, you mutate it which creates data reactivity and async gotchas.
```

> - New functionality. There are computed properties, watchers, mixins, etc that don't exist in React out of the box

Reviewer: ```suggestion
- New constructs. There are computed properties, watchers, and mixins that don't exist in React out of the box
```

> - You write HTML and styles the way you would traditionally in a regular index.html page

Reviewer: ```suggestion
- You write HTML and styles the way you would traditionally in a regular HTML page
```

> - PropTypes are enforced in Vue, it's optional in React

Reviewer: Are they called PropTypes in Vue as well?

Author: Yeah it's called Prop Types as well
